### ISSUE TITLE
How to set printf buffer and direct to debug callback

### ISSUE THREAD

**Issue by seanbaxter** [KhronosGroup/Vulkan-ValidationLayers#2464]
How do I programmatically set the printf buffer size? It's way too small by default. I want to set the buffer size and use it without the validation layer turned on, as the doc suggests.

Also, I only see output when I export DEBUG_PRINTF_TO_STDOUT. I do have a VkDebugUtilsMessengerCreateInfoEXT user callback registered, but it isn't getting fired with the printf text. I am setting the INFO severity bit. I do get the usual Vulkan errors on this callback.

My compute shader is here:
https://github.com/seanbaxter/vulkan_compute/blob/master/compute.cxx#L8

The setup code is here:
https://github.com/seanbaxter/vulkan_compute/blob/master/context.cxx

This is my first Vulkan program and uhh. I don't know what I'm doing. Surprised I got this far.

**Comment by mark-lunarg** [comment]
@Tony-LunarG, hard to tell, but this looks like GPU-AV. Is there a way to set the buffer size?  Also @seanbaxter, can you clarify "..use it without the validation layer turned on, as the doc suggests"?

**Comment by seanbaxter** [comment]
"Validation itself is not necessary for Debug Printf and can be disabled without affecting Debug Printf functionality." I found the vkconfig tools with all the checkboxes, so I think that's what this refers to. But this is backburner until the vkCreateShaderModule hangs are resolved.

**Comment by mark-lunarg** [comment]
@seanbaxter, thanks!  @Tony-LunarG, 'validation' seems like it could be a bit misleading, since debug-printf is part of 'KHRONOS_validation'.  Just sayin'.

**Comment by Tony-LunarG** [comment]
Yes, you have to enable the KHRONOS_validation layer, but it is advised that you turn off the standard validation checks.  vkconfig is the easiest way to do so.  Selecting debug printf in vkconfig should disable the standard validation checks automatically.
Right now, the only way to set the debug printf buffer size is through a vk_layer_settings.txt file.  Set khronos_validation.printf_buffer_size = <desired size of buffer>  Note that the printf buffers are in device memory and that they are a per-draw resource, so it can eat up memory somewhat quickly if you aren't careful.  Starting with the next SDK, you should be able to use vkconfig to configure the debug printf buffer size.
When you say you have a VkDebugUtilsMessengerCreateInfoEXT user callback registered, did you register it both in the CreateInstance pNext chain and then again after CreateInstance in a call to vkCreateDebugUtilsMessengerEXT?

**Comment by seanbaxter** [comment]
Ya it's registered in both those places. After fiddling with vkconfig, and having INFO debug messenger severity, I get printfs formatted to stdout as validator info, eg:
```
[INFO: Validation]
Validation Information: [ UNASSIGNED-DEBUG-PRINTF ] Object 0: handle = 0x255e2e8, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x92394c89 | thread = 154
```

When I export `DEBUG_PRINTF_TO_STDOUT` I see them as regular messages:
`thread = 154`

Don't know if this is by design. Doesn't matter that much though.

**Comment by Tony-LunarG** [comment]
That looks like your callback isn't being called.  When you don't have a callback, validation and debug printf messages go to stdout, and debug printf will call LogInfo and that adds the validation formatting.  When you export DEBUG_PRINTF_TO_STDOUT, debug printf does not call LogInfo and instead just sends the message to stdout directly.

**Comment by Tony-LunarG** [comment]
Closing assuming you got it working.  Re-open if that's not the case

**Comment by Inori** [comment]
Sorry to reopen this.

`khronos_validation.printf_to_stdout` seems not work on my machine.

The validation messages shows that I'm using the right settings file and other options works right (like `khronos_validation.enables`).

Only the `printf_to_stdout` seems doesn't make sense.
Weither I set it to `true` or `false`, I got debug messages from both debug callback and stdout.

Vulkan SDK version: 1.3.211.0
